pkg/appgateserver/metrics.go

```diff
@@ -0,0 +1,36 @@
+package appgateserver
+
+import (
+	"github.com/go-kit/kit/metrics"
+	"github.com/go-kit/kit/metrics/prometheus"
+	stdprometheus "github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	// relaysTotal is a counter metric that represents the total number of requests processed by the AppGate server.
+	// It is a Counter type metric, which means it is used to accumulate counts of events (in this case, the total number of requests).
+	// This metric is crucial for understanding the overall workload and traffic handled by the server.
+	// It is labeled by 'service_id', which allows for distinguishing and aggregating request counts across different services managed by the AppGate server.
+	//
+	// Example of usage:
+	// - Monitoring the aggregate load and request rate over time.
+	// - Comparing request volumes across different services managed by the AppGate server.
+	relaysTotal metrics.Counter = prometheus.NewCounterFrom(stdprometheus.CounterOpts{
+		Subsystem: "appgateserver",
+		Name:      "requests_total",
+	}, []string{"service_id"})
+
+	// relaysErrorsTotal is a counter metric that tracks the total number of errors encountered by the AppGate server.
+	// This metric increments each time an error occurs, providing insight into the health and stability of the server.
+	// Tracking error rates is essential for maintaining the reliability of the server and for identifying issues that require attention.
+	// The metric is labeled by 'service_id', which allows for tracking and analyzing error rates per service,
+	// enabling targeted troubleshooting and performance optimization.
+	//
+	// Example of usage:
+	// - Monitoring the overall health and error rates of the server.
+	// - Identifying services with higher error rates for targeted debugging and improvement.
+	relaysErrorsTotal metrics.Counter = prometheus.NewCounterFrom(stdprometheus.CounterOpts{
+		Subsystem: "appgateserver",
+		Name:      "errors_total",
+	}, []string{"service_id"})
+)
```

Comment:
I would also add a duration metric for the `AppGateServer`, operators would want to have that info too.

Reply:
There's a library I hook to HTTP server via a Middleware, and it provides these metrics:
```
http_request_duration_seconds_bucket{code="200",handler="/anvil",method="POST",service="",le="0.005"} 0
http_request_duration_seconds_bucket{code="200",handler="/anvil",method="POST",service="",le="0.01"} 0
http_request_duration_seconds_bucket{code="200",handler="/anvil",method="POST",service="",le="0.025"} 5
http_request_duration_seconds_bucket{code="200",handler="/anvil",method="POST",service="",le="0.05"} 5
http_request_duration_seconds_bucket{code="200",handler="/anvil",method="POST",service="",le="0.1"} 6
http_request_duration_seconds_bucket{code="200",handler="/anvil",method="POST",service="",le="0.25"} 6
http_request_duration_seconds_bucket{code="200",handler="/anvil",method="POST",service="",le="0.5"} 6
http_request_duration_seconds_bucket{code="200",handler="/anvil",method="POST",service="",le="1"} 6
http_request_duration_seconds_bucket{code="200",handler="/anvil",method="POST",service="",le="2.5"} 6
http_request_duration_seconds_bucket{code="200",handler="/anvil",method="POST",service="",le="5"} 6
http_request_duration_seconds_bucket{code="200",handler="/anvil",method="POST",service="",le="10"} 6
http_request_duration_seconds_bucket{code="200",handler="/anvil",method="POST",service="",le="+Inf"} 6
http_request_duration_seconds_sum{code="200",handler="/anvil",method="POST",service=""} 0.17413220799999998
http_request_duration_seconds_count{code="200",handler="/anvil",method="POST",service=""} 6
# HELP http_requests_inflight The number of inflight requests being handled at the same time.
# TYPE http_requests_inflight gauge
http_requests_inflight{handler="/anvil",service=""} 0
# HELP http_response_size_bytes The size of the HTTP responses.
# TYPE http_response_size_bytes histogram
http_response_size_bytes_bucket{code="200",handler="/anvil",method="POST",service="",le="100"} 6
http_response_size_bytes_bucket{code="200",handler="/anvil",method="POST",service="",le="1000"} 6
http_response_size_bytes_bucket{code="200",handler="/anvil",method="POST",service="",le="10000"} 6
http_response_size_bytes_bucket{code="200",handler="/anvil",method="POST",service="",le="100000"} 6
http_response_size_bytes_bucket{code="200",handler="/anvil",method="POST",service="",le="1e+06"} 6
http_response_size_bytes_bucket{code="200",handler="/anvil",method="POST",service="",le="1e+07"} 6
http_response_size_bytes_bucket{code="200",handler="/anvil",method="POST",service="",le="1e+08"} 6
http_response_size_bytes_bucket{code="200",handler="/anvil",method="POST",service="",le="1e+09"} 6
http_response_size_bytes_bucket{code="200",handler="/anvil",method="POST",service="",le="+Inf"} 6
http_response_size_bytes_sum{code="200",handler="/anvil",method="POST",service=""} 240
http_response_size_bytes_count{code="200",handler="/anvil",method="POST",service=""} 6
```

PRO: it's working and just a few lines change.
CON: I can't change metric names/labels. I feel like "http" doesn't really capture what we want to do in AppGate Server because there are other use cases where traffic is sent via tcp (and potentially udp in future). We can build our own implementation, but that would require adding a custom writer. I didn't feel comfortable doing that, but this is probably the way to go long term, in the future. If you feel we should do that — I'm happy to try! :)